securedrop/alembic/versions/17c559a7a685_pgp_public_keys.py

```diff
@@ -0,0 +1,92 @@
+"""PGP public keys
+
+Revision ID: 17c559a7a685
+Revises: 811334d7105f
+Create Date: 2023-09-21 12:33:56.550634
+
+"""
+
+import traceback
+
+import pretty_bad_protocol as gnupg
+import sqlalchemy as sa
+from alembic import op
+from encryption import EncryptionManager
+from sdconfig import SecureDropConfig
+
+# revision identifiers, used by Alembic.
+revision = "17c559a7a685"
+down_revision = "811334d7105f"
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    """
+    Migrate public keys from the GPG keyring into the SQLite database
+
+    We iterate over all the secret keys in the keyring, identify the
+    corresponding source, export the key and save it in the database.
```

Comment:
SGTM!

Yeah, basically I was thinking of some case where the source was deleted out of the database and deletion out of the GPG keyring failed.

Reply:
Re-reading prime_keycache reminded me why I did it this way. And that's because it's impossible to do it any other way with our current GPG wrappers/abstractions.

If you look at prime_keycache, it does:
```python
        for source in Source.query.filter_by(pending=False, deleted_at=None).all():
            try:
                encryption_mgr.get_source_public_key(source.filesystem_id)
            except GpgKeyNotFoundError:
                pass
```

Which you think means its iterating over each source, and looking up the public key for each. But EncryptionManager._get_source_key_details() is iterating over the keyring each time before finding a matching key, (aka it's O(N²)). So we just iterate over the keyring once and matching sources / exporting those keys (O(N)).

Reply:
Yeah, that was my intuition.  Thanks for confirming it based on the current implementation.